internal/validation/validation.go

```diff
@@ -0,0 +1,290 @@
+package validation
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
+	"github.com/openfga/openfga/storage"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+)
+
+// ValidateTuple checks whether a tuple is welformed and valid according to the provided model.
+func ValidateTuple(typesys *typesystem.TypeSystem, tk *openfgapb.TupleKey) error {
+
+	if err := ValidateUser(typesys, tk); err != nil {
+		return err
+	}
+
+	if err := ValidateObject(typesys, tk); err != nil {
+		return err
+	}
+
+	if err := ValidateRelation(typesys, tk); err != nil {
+		return err
+	}
+
+	// now we assume our tuple is well-formed, it's time to check
+	// the tuple against other model and type-restriction constraints
+
+	err := validateTuplesetRestrictions(typesys, tk)
+	if err != nil {
+		return err
+	}
+
+	objectType := tuple.GetType(tk.GetObject())
+	relation := tk.GetRelation()
+
+	hasTypeInfo, err := typesys.HasTypeInfo(objectType, relation)
+	if err != nil {
+		return err
+	}
+
+	if hasTypeInfo {
+		err := validateTypeRestrictions(typesys, tk)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// validateTuplesetRestrictions validates the provided TupleKey against tupleset restrictions.
+//
+// Given a rewrite definition such as 'viewer from parent', the 'parent' relation is known as the
+// tupleset. This method ensures the following are *not* possible:
+//
+// 1. `document:1#parent@folder:1#parent` (cannot evaluate/assign a userset value to a tupleset relation)
+// 2. `document:1#parent@*` (cannot evaluate/assign untyped wildcard to a tupleset relation (1.0 models))
+// 3. `document:1#parent@folder:*` (cannot evaluate/assign typed wildcard to a tupleset relation (1.1. models))
+func validateTuplesetRestrictions(typesys *typesystem.TypeSystem, tk *openfgapb.TupleKey) error {
+
+	objectType := tuple.GetType(tk.GetObject())
+	relation := tk.GetRelation()
+
+	isTupleset, err := typesys.IsTuplesetRelation(objectType, relation)
+	if err != nil {
+		return err
+	}
+
+	if !isTupleset {
+		return nil
+	}
+
+	rel, err := typesys.GetRelation(objectType, relation)
+	if err != nil {
+		return err
+	}
+
+	rewrite := rel.GetRewrite().GetUserset()
+
+	// tupleset relation involving a rewrite
+	if rewrite != nil && reflect.TypeOf(rewrite) != reflect.TypeOf(&openfgapb.Userset_This{}) {
+		return &tuple.InvalidTupleError{
+			Reason:   fmt.Sprintf("unexpected rewrite encountered with tupelset relation '%s#%s'", objectType, relation),
+			TupleKey: tk,
+		}
+	}
```

Comment:
Hmmm does this code belong here? It is not validating the tuple, but the rewrite itself

Reply:
I agree, ideally it wouldn't belong here. The reason it does is because we may still be encountering developers that were victims of the recent bug reports that may have models involving rewrites on tupleset relations. At least this consolidates the validation into one place instead of having to exist in Expand, Check, etc.. individually. 

It'd be nice to remove it in favor of these validations at the `WriteAuthorizationModel` API.

Reply:
Ok so we can leave this code for a while and we'll remove it.. when? Maybe leave a TODO?

Reply:
That's the million dollar question.. I'm not sure when is best to remove this code. I think that we should just bite the bullet and remove it as soon as we have verified (if we haven't already) that `WriteAuthorizationModel` RPC no longer allows models with rewrites on tupleset relations.

Reply:
> WriteAuthorizationModel RPC no longer allows models with rewrites on tupleset relations.

Yes, we have that: https://github.com/openfga/openfga/blob/57d9b3da62c4634a73c47d839c6be7ac0e0b6dde/pkg/typesystem/typesystem.go#L532-L537